In [1]:
import numpy as np
import pandas as pd

# Identify:

Please fill-in your ID in the following cell, **only id numbers, no names!**


In [2]:
id1 = "318465291"
id2 = "123456789"

In [3]:
df = pd.DataFrame([id1])

Uncomment in case two students are doing the assignment together:

In [4]:
#df = pd.DataFrame([id1,id2])

In [5]:
df.to_clipboard(index=False,header=False)

Please fill your names here:

Daniel tugendhaft

# Q1

In [6]:
url = 'https://raw.githubusercontent.com/nlihin/data-analytics/main/datasets/aircraft%20wildlife%20strikes.csv'
aws_df = pd.read_csv(url)

/var/folders/1n/l8xnr__x38x4jqs4jb3_b1_w0000gp/T/ipykernel_18758/105516896.py:2: DtypeWarning: Columns (6,12,14) have mixed types. Specify dtype option on import or set low_memory=False.
  aws_df = pd.read_csv(url)


In [7]:
aws_df[["Airport","Height","Speed","Species Quantity","Injuries"]].isnull().sum().to_frame('nulls')

nulls
Airport              290
Height             59749
Speed              84565
Species Quantity    4125
Injuries          134882

# Q2

In [8]:
aws_df["Species Quantity"] = aws_df["Species Quantity"].fillna('unreported')

In [9]:
aws_df["Species Quantity"].unique()

array(['1', '02-Oct', 'unreported', '11-100', 'Over 100'], dtype=object)

# Q3

In [10]:
SpeciesQuantity_counts = pd.DataFrame(aws_df[["Species Quantity"]].value_counts(), columns = ['Quantity'])
SpeciesQuantity_counts

Quantity
Species Quantity          
1                   114880
02-Oct               15187
unreported            4125
11-100                 794
Over 100                22

In [11]:
SpeciesQuantity_counts.reset_index()

Species Quantity  Quantity
0                1    114880
1           02-Oct     15187
2       unreported      4125
3           11-100       794
4         Over 100        22

In [12]:
aws_df["Species Quantity"] = aws_df["Species Quantity"].replace('02-Oct','2-10')
pd.DataFrame(aws_df[["Species Quantity"]].value_counts(), columns = ['Quantity']).reset_index()

Species Quantity  Quantity
0                1    114880
1             2-10     15187
2       unreported      4125
3           11-100       794
4         Over 100        22

# Q4

In [13]:
aws_df['Aircraft Type'].isna().value_counts()

False    98403
True     36605
Name: Aircraft Type, dtype: int64

There are 36605 empty values in the column - 'Aircraft Type'

# Q5

In [14]:
aws_df.shape

(135008, 60)

In [15]:
aws_df.dropna(axis=1,thresh =len(aws_df)-30000).shape

(135008, 38)

The number of columns dropped from the table are 22.

# Q6

In [16]:
aws_df[['Height_new']] = aws_df[['Height']].fillna(aws_df[['Height']].median())

In [17]:
aws_df[['Height','Height_new']]

Height  Height_new
0          NaN        50.0
1          0.0         0.0
2          NaN        50.0
3          NaN        50.0
4          NaN        50.0
...        ...         ...
135003     NaN        50.0
135004  5000.0      5000.0
135005  3000.0      3000.0
135006  6000.0      6000.0
135007     NaN        50.0

[135008 rows x 2 columns]

# Q7

In [18]:
aws_YearSpecies = aws_df.groupby(['Incident Year'])[['Species Quantity']].value_counts().unstack()
aws_YearSpecies = aws_YearSpecies.reset_index()
aws_YearSpecies = aws_YearSpecies.rename_axis(None, axis=1)
aws_YearSpecies

Incident Year        1  11-100    2-10  Over 100  unreported
0            2001   5257.0    51.0   835.0       3.0        70.0
1            2002   5652.0    51.0   905.0       2.0        49.0
2            2003   5448.0    53.0   896.0       3.0        85.0
3            2004   5636.0    52.0   858.0       NaN       576.0
4            2005   6298.0    60.0   861.0       1.0       371.0
5            2006   6270.0    68.0   889.0       2.0       521.0
6            2007   6684.0    45.0  1000.0       NaN       264.0
7            2008   6653.0    51.0   910.0       2.0       377.0
8            2009   8297.0    49.0  1144.0       NaN       358.0
9            2010   8644.0    49.0  1198.0       1.0        42.0
10           2011   8865.0    61.0  1171.0       1.0       265.0
11           2012   9721.0    51.0  1115.0       2.0       375.0
12           2013  10166.0    49.0  1162.0       2.0       287.0
13           2014  12216.0    60.0  1378.0       1.0       321.0
14           2015   9073.0    44.0   865.0       2.0       164.0

# Q8

The meaning of the "NaN" is that in the years - 2004, 2007 and 2009 the Species Quantity was no over 100 in any incident.

# Q9

In [19]:
aws_HeightSpeed = aws_df.groupby(['Incident Year','Incident Month']).agg({'Height':'mean','Speed':'median'})
aws_HeightSpeed = aws_HeightSpeed.reset_index()
aws_HeightSpeed

Incident Year  Incident Month       Height  Speed
0             2001               1   562.480769  130.0
1             2001               2   647.482993  135.0
2             2001               3  1152.746479  140.0
3             2001               4  1401.976271  140.0
4             2001               5  1106.207386  137.0
..             ...             ...          ...    ...
172           2015               5  1210.496350  150.0
173           2015               6   318.896484  130.0
174           2015               7   355.818999  130.0
175           2015               8   683.487691  140.0
176           2015               9  1406.957908  143.0

[177 rows x 4 columns]